File: scripts/hygiene/divergence-enum-coverage.py
#!/usr/bin/env python3
# make divergence-enum-coverage — issue #36's third meta-test: "divergence
# enum values no fixture exercises" (Article V.3).
#
# Known edge case in parseEnumMembers() below, not a bug to silently work
# around: it splits the enum body on "," up front, before stripping "//" /
# "/* */" comments out of each resulting piece — a comment containing a
# literal comma (or a block comment spanning a comma) is split apart first
# and only has its comment markers stripped from each half separately.
# Tightening that parsing edge case is a separate concern from this scan's
# own job (declared-divergence coverage) and is left alone here.
import os
import re
import sys

args = sys.argv[1:]
root = args[args.index("--root") + 1] if "--root" in args else os.getcwd()

SRC_DIR = os.path.join(root, "src")
FIXTURES_DIR = os.path.join(root, "test", "fixtures")
SQL_TEST_DIR = os.path.join(root, "test", "sql")


def list_files(dir_, pred):
    if not os.path.isdir(dir_):
        return []
    out = []

    def walk(d):
        for entry in os.listdir(d):
            full = os.path.join(d, entry)
            if os.path.isdir(full):
                walk(full)
            elif pred(entry):
                out.append(full)

    walk(dir_)
    return out


# `enum class <Name>Divergence { ... }` — a flat, comment-bearing C++ enum.
# No nested braces are expected inside a declared-divergence enum body, so a
# non-greedy match up to the first `}` is sufficient.
DIVERGENCE_ENUM_RE = re.compile(r"enum\s+class\s+([A-Za-z_][A-Za-z0-9_]*Divergence)\s*\{([\s\S]*?)\}")


def parse_enum_members(body):
    result = []
    for part in body.split(","):
        part = re.sub(r"//.*$", "", part, flags=re.MULTILINE)
        part = re.sub(r"/\*[\s\S]*?\*/", "", part)
        part = part.strip()
        if not part:
            continue
        name = part.split("=")[0].strip()
        if re.fullmatch(r"[A-Za-z_][A-Za-z0-9_]*", name):
            result.append(name)
    return result


def find_declared_divergences(root):
    enums = []  # {enumName, member, file}
    for file in list_files(SRC_DIR, lambda f: bool(re.search(r"\.(hpp|hh|h|cpp|cc|cxx)$", f))):
        try:
            with open(file, "rb") as fh:
                raw = fh.read()
        except OSError:
            continue
        if b"\0" in raw:
            continue
        content = raw.decode("utf-8", errors="replace")
        for m in DIVERGENCE_ENUM_RE.finditer(content):
            for member in parse_enum_members(m.group(2)):
                enums.append({"enumName": m.group(1), "member": member, "file": file[len(root) + 1:]})
    return enums


def is_exercised(member, corpus):
    return any(member in text for text in corpus)


declared = find_declared_divergences(root)
corpus_files = list_files(FIXTURES_DIR, lambda f: f.endswith(".yaml") or f.endswith(".yml")) + list_files(
    SQL_TEST_DIR, lambda f: f.endswith(".test")
)
corpus = []
for f in corpus_files:
    try:
        with open(f, "r", encoding="utf-8", errors="replace") as fh:
            corpus.append(fh.read())
    except OSError:
        corpus.append("")

violations = []
for d in declared:
    if not is_exercised(d["member"], corpus):
        violations.append(f'{d["file"]}: {d["enumName"]}::{d["member"]} is a declared divergence no fixture or sqllogictest exercises')

if violations:
    print("divergence-enum-coverage: FAIL", file=sys.stderr)
    for v in violations:
        print(f"  {v}", file=sys.stderr)
    sys.exit(1)
print(f"divergence-enum-coverage: PASS ({len(declared)} declared divergence(s))")

File: scripts/hygiene/divergence-enum-coverage.mjs
#!/usr/bin/env bun
// make divergence-enum-coverage — issue #36's third meta-test: "divergence
// enum values no fixture exercises" (Article V.3: "The only sanctioned
// exclusion is a declared divergence in the comparator's closed enum; an
// enum value no fixture exercises is a failure."; T13 and
// `docs/testing/live-oracles.md`'s "Declared divergences" section: "the L13
// meta-test fails if an enum value is never exercised by any fixture — an
// unused divergence is a tolerance waiting to be used.").
//
// `docs/testing/live-oracles.md` shows the closed enums this rule will
// eventually govern (`ChdbDivergence`, `TimescaleDivergence`,
// `ChplanDivergence`) as a design sketch for the L6a live-oracle harness —
// which is not part of M1 (`docs/testing/layers.md`'s L6a row is [MERGE]
// for chDB and [RELEASE] for Timescale, and M1's own milestone description
// lists only L0, L1a/b, L2, L4, L5, L13). Building that harness is out of
// this issue's scope. What this scan fences instead is the mechanism
// itself, against wherever a real `enum class *Divergence` actually lives
// in the kernel's own source — so the rule is already live and enforced the
// day a future issue adds the first one, rather than a second thing that
// issue would also need to remember to wire up. Today `src/` declares no
// such enum, so this scan is vacuously green on the real tree; its fixture
// branch proves the mechanism on a synthetic one (Article V.5 still needs a
// real, provably-executed lane — `scripts/hygiene.mjs`'s TREE_SCANS — not a
// hope that a future issue remembers this file exists).
import { readFileSync, readdirSync, statSync, existsSync } from "node:fs";
import { join } from "node:path";

const args = process.argv.slice(2);
const rootIdx = args.indexOf("--root");
const root = rootIdx === -1 ? process.cwd() : args[rootIdx + 1];

const SRC_DIR = join(root, "src");
const FIXTURES_DIR = join(root, "test", "fixtures");
const SQL_TEST_DIR = join(root, "test", "sql");

function listFiles(dir, pred) {
  if (!existsSync(dir)) return [];
  const out = [];
  const walk = (d) => {
    for (const entry of readdirSync(d)) {
      const full = join(d, entry);
      const st = statSync(full);
      if (st.isDirectory()) walk(full);
      else if (pred(entry)) out.push(full);
    }
  };
  walk(dir);
  return out;
}

// `enum class <Name>Divergence { ... }` — a flat, comment-bearing C++ enum,
// the shape `docs/testing/live-oracles.md`'s own shown block uses. No
// nested braces are expected inside a declared-divergence enum body, so a
// non-greedy match up to the first `}` is sufficient (the same
// "regex scan, not a real parser" posture `registry-closure.mjs` documents).
const DIVERGENCE_ENUM_RE = /enum\s+class\s+([A-Za-z_][A-Za-z0-9_]*Divergence)\s*\{([\s\S]*?)\}/g;

function parseEnumMembers(body) {
  return body
    .split(",")
    .map((part) => part.replace(/\/\/.*$/gm, "").replace(/\/\*[\s\S]*?\*\//g, "").trim())
    .filter(Boolean)
    .map((part) => part.split("=")[0].trim()) // drop an explicit initializer, if any
    .filter((name) => /^[A-Za-z_][A-Za-z0-9_]*$/.test(name));
}

function findDeclaredDivergences(root) {
  const enums = []; // { enumName, member, file }
  for (const file of listFiles(SRC_DIR, (f) => /\.(hpp|hh|h|cpp|cc|cxx)$/.test(f))) {
    let content;
    try {
      content = readFileSync(file, "utf8");
    } catch {
      continue;
    }
    if (content.includes("\0")) continue;
    DIVERGENCE_ENUM_RE.lastIndex = 0;
    let m;
    while ((m = DIVERGENCE_ENUM_RE.exec(content))) {
      for (const member of parseEnumMembers(m[2])) {
        enums.push({ enumName: m[1], member, file: file.slice(root.length + 1) });
      }
    }
  }
  return enums;
}

// "Exercised by a fixture" is deliberately textual, not schema-shaped: the
// fixture format (`test/fixtures/schema.json`) has no `divergence:` field of
// its own to add — doing so is a fixture-format change no future L6a issue
// has asked for yet, and Article III.4 rules out pre-building tooling a
// future issue owns. A fixture "exercises" a declared divergence by naming
// it — in its `fixture:` id, a `wrong:` key, a comment, or any other text —
// anywhere in the fixture file itself, or in the sqllogictest that drives
// the row the enum concerns.
function isExercised(member, corpus) {
  return corpus.some((text) => text.includes(member));
}

const declared = findDeclaredDivergences(root);
const corpusFiles = [...listFiles(FIXTURES_DIR, (f) => f.endsWith(".yaml") || f.endsWith(".yml")), ...listFiles(SQL_TEST_DIR, (f) => f.endsWith(".test"))];
const corpus = corpusFiles.map((f) => {
  try {
    return readFileSync(f, "utf8");
  } catch {
    return "";
  }
});

const violations = [];
for (const { enumName, member, file } of declared) {
  if (!isExercised(member, corpus)) {
    violations.push(`${file}: ${enumName}::${member} is a declared divergence no fixture or sqllogictest exercises`);
  }
}

if (violations.length > 0) {
  console.error("divergence-enum-coverage: FAIL");
  for (const v of violations) console.error(`  ${v}`);
  process.exit(1);
}
console.log(`divergence-enum-coverage: PASS (${declared.length} declared divergence(s))`);
